core/utils/STree.py

```diff
@@ -55,20 +55,64 @@ def compute_s_link(s: str, u: Node) -> None:
     u.s_link = v
 
 
+def traverse_edge(text: str, pattern: str, idx: int, start: int, end: int) -> tuple[int, int]:
+    i: int = start
+
+    while i < end and idx < len(pattern):
+        if text[i] != pattern[idx]:
+            return -1, idx
+
+        i += 1
+        idx += 1
+
+    # we have a match
+    if idx == len(pattern):
+        return 1, idx
+
+    # 0 means there are still more characters to match
+    return 0, idx
+
+
+def perform_search(node: Node, text: str, pattern: str, idx: int) -> int:
+    # case where the pattern starts with same alphabet as text
+    if node.idx == 0 and node.depth >= len(pattern):
+        return text.startswith(pattern)
+
+    if node.idx != 0:
+
+        """
+        We will traverse the edge and match alphabet by alphabet
```

Comment:
This should probably be a comment and not a string. It doesn't function like a documentation string when it is in the middle of a function.

Reply:
My bad